File: AspNetMvcUnleashed/0672329980_source/Chapter05/VB/MvcApplication1/MvcApplication1/Models/ProductRepository.vb
Public Class ProductRepository
	Implements IProductRepository

	Private _entities As New ProductsDBEntities()

	#Region "IProductRepository Members"

	Public Function List() As IEnumerable(Of Product) Implements IProductRepository.List
		Return _entities.ProductSet.ToList()
	End Function

	Public Function [Get](ByVal id As Integer) As Product Implements IProductRepository.Get
		Return (From p In _entities.ProductSet _
		        Where p.Id = id _
		        Select p).FirstOrDefault()
	End Function

	Public Sub Create(ByVal productToCreate As Product) Implements IProductRepository.Create
		_entities.AddToProductSet(productToCreate)
		_entities.SaveChanges()
	End Sub

	Public Sub Edit(ByVal productToEdit As Product) Implements IProductRepository.Edit
		Dim originalProduct = [Get](productToEdit.Id)
		_entities.ApplyPropertyChanges(originalProduct.EntityKey.EntitySetName, productToEdit)
		_entities.SaveChanges()

	End Sub

	Public Sub Delete(ByVal productToDelete As Product) Implements IProductRepository.Delete 
		Dim originalProduct = [Get](productToDelete.Id)
		_entities.DeleteObject(originalProduct)
		_entities.SaveChanges()
	End Sub

	#End Region

End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter05/CS/MvcApplication1/MvcApplication1/Models/ProductRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace MvcApplication1.Models
{
    public class ProductRepository : IProductRepository
    {
        private ProductsDBEntities _entities = new ProductsDBEntities();

        #region IProductRepository Members

        public IEnumerable<Product> List()
        {
            return _entities.ProductSet.ToList();
        }

        public Product Get(int id)
        {
            return (from p in _entities.ProductSet
                          where p.Id == id
                          select p).FirstOrDefault();
        }

        public void Create(Product productToCreate)
        {
            _entities.AddToProductSet(productToCreate);
            _entities.SaveChanges();
        }

        public void Edit(Product productToEdit)
        {
            var originalProduct = Get(productToEdit.Id);
            _entities.ApplyPropertyChanges(originalProduct.EntityKey.EntitySetName, productToEdit);
            _entities.SaveChanges();

        }

        public void Delete(Product productToDelete)
        {
            var originalProduct = Get(productToDelete.Id);
            _entities.DeleteObject(originalProduct);
            _entities.SaveChanges();
        }

        #endregion
    }
}
